### PR TITLE
Enhance Boot Flow documentation

### PR DESCRIPTION
**Pull request by atharvalele**
- Add details about each stage and what each stage does
- Add details about what data is passed and how it is passed between stages
- Add information about Loader Global Data and how it is used by the bootloader

### CODE REVIEW

File: source/developer-guides/boot-flow.rst [hunk 1]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
   If verification fails, one of two things occurs. If resiliency is enabled, the boot is swapped over to the alternate partition
   to recover the broken partition (see :ref:`firmware-resiliency-and-recovery`). Otherwise, the boot is halted.
 
+Stage 1A
+~~~~~~~~
+
+When coming out of reset, the processor may be in a very raw state and needs some basic
+initialization before the rest of the platform can be initialized. Stage 1A sets up the
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:43]: Maybe rephrase as - "After coming out of reset, Stage 1A sets up the ...

File: source/developer-guides/boot-flow.rst [hunk 2]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
   If verification fails, one of two things occurs. If resiliency is enabled, the boot is swapped over to the alternate partition
   to recover the broken partition (see :ref:`firmware-resiliency-and-recovery`). Otherwise, the boot is halted.
 
+Stage 1A
+~~~~~~~~
+
+When coming out of reset, the processor may be in a very raw state and needs some basic
+initialization before the rest of the platform can be initialized. Stage 1A sets up the
+initial exectution environment to support the execution of the following stages. At this
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:44]: typo - execution

File: source/developer-guides/boot-flow.rst [hunk 3]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
   If verification fails, one of two things occurs. If resiliency is enabled, the boot is swapped over to the alternate partition
   to recover the broken partition (see :ref:`firmware-resiliency-and-recovery`). Otherwise, the boot is halted.
 
+Stage 1A
+~~~~~~~~
+
+When coming out of reset, the processor may be in a very raw state and needs some basic
+initialization before the rest of the platform can be initialized. Stage 1A sets up the
+initial exectution environment to support the execution of the following stages. At this
+stage, a stack is not available and hence assembly used to carry out the basic initialization
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:45]: After coming out of reset, when the processor starts executing from the reset vector, there may not be writeable memory and a stack might have not been setup. So, the reset vector code is typically implemented using assembly language instructions. The lack of writeable memory also prevents usage of global variables.

File: source/developer-guides/boot-flow.rst [hunk 4]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
 
+Stage 1A
+~~~~~~~~
+
+When coming out of reset, the processor may be in a very raw state and needs some basic
+initialization before the rest of the platform can be initialized. Stage 1A sets up the
+initial exectution environment to support the execution of the following stages. At this
+stage, a stack is not available and hence assembly used to carry out the basic initialization
+steps. Stage 1A is also excuted from the flash (Read-only memory) and thus does not use
+global variables.
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:47]: The lack of writeable memory also prevents usage of global variables.

File: source/developer-guides/boot-flow.rst [hunk 5]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+
+For Intel x86 processors, the first few instructions for the reset vector code need to be
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:49]: Remove the line from "For Intel...CPU to begin execution"

File: source/developer-guides/boot-flow.rst [hunk 6]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+at the predefined reset vector address for the CPU to begin execution. Stage 1A contains a
+module called VTF (Volume Top File) which is placed at the top within the Stage 1A FV.
+The VTF contains the reset vector code and hence the VTF needs to be placed at an appropriate
+address so that the reset vector matches the Vtf0 file (Identified by the GUID ``1BA0062E-C779-4582-8566-336AE8F78F09``).
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:53]: placed at an appropriate address so that the reset vector code in the Vtf0 file aligns with the reset vector of Intel x86 architecture.

File: source/developer-guides/boot-flow.rst [hunk 7]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+The entry point address is specified by the ``_ModuleEntryPoint`` symbol is placed as the first
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:54]: The entry point for the Stage1A module within the Stage1A FV (_ModuleEntryPoint)  is placed as tge furst DWORD...

File: source/developer-guides/boot-flow.rst [hunk 8]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+DWORD of the built FV. The reset vector code from the Vtf0 jumps to this address and continues
+from the ``_ModuleEntryPoint`` defined in ``SecEntry.nasm``.
+
+After coming out of reset, Stage 1A performs the following initialization steps:
+
+* Switch from 16-bit real mode to 32-bit protected mode to provide full 32-bit memory addressibility
+* Calls the FSP-T provided ``TempRamInit()`` API which sets up Cache-As-RAM (CAR)
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:61]: FSP-T provided ``TempRamInit()`` API which configures a temporary writeable memory

File: source/developer-guides/boot-flow.rst [hunk 9]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+* Sets up a stack in the CAR region
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:62]: CAR -> Temporary memory initialized by FSP's TempRamInit API

File: source/developer-guides/boot-flow.rst [hunk 10]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+* ``SecEntry.nasm`` pushes the elements of the struct on the stack in the expected order
+
+**Data Passed from Stage 1A to Stage 1B** --
+
+* ``Stage1A_PARAM`` struct
+
+Stage 1B
+~~~~~~~~
+
+The primary purpose of Stage1B is to bring up the system memory and the stage performs the following
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:82]: system memory and performs the following

File: source/developer-guides/boot-flow.rst [hunk 11]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+initialization steps:
+
+* Early platform init
+* Load and build configuration database
+* Calls FSP-M provided ``FspMemoryInit()`` which initializes the system's main memory (RAM)
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:87]: remove (RAM)

File: source/developer-guides/boot-flow.rst [hunk 12]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+* Sets up the global data structures in RAM and migrates LdrGlobal from CAR to RAM
+* Saves system memory information like - TOLUM (Top of low usable memory), TOUUM (Top of upper usable memory), etc.
+* Switches from CAR stack to RAM stack
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:90]: CAR --> Stack in temporary memory to one in the main memory

File: source/developer-guides/boot-flow.rst [hunk 13]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+
+Data Passed from Stage 1B to Stage 2
+
+* ``STAGE2_PARAM`` struct
+
+Stage 2
+~~~~~~~
+
+Stage 2 is the "post-memory" stage and is responsible for completing system initialization after
+main memory is available. Stage 2 code is run from RAM and may be compressed. Stage 2 includes
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:105]: RAM --> main memory

File: source/developer-guides/boot-flow.rst [hunk 14]
```diff
@@ -36,6 +36,118 @@ Payload    Load, verify and launch OS images; or perform firmware update
...
+* Prepares HOBs required by the payload
+* Calls FSP-S provided ``FspNotifyPhase`` APIs - ``ReadyToBoot``, ``EndOfFirmware``
+* Loads and jumps to the payload
+
+
+LdrGlobal - Loader Global Data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The Loader Global Data structure is used to store important information throughout the
+lifecycle of the bootloader. It consists of information like available memory start and
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:123]: throughout the different execution phases of the bootloader

**Comment by atharvalele** [comment]
Thanks for the suggestions @rprangar . I've updated the PR.

File: source/developer-guides/boot-flow.rst [hunk 15]
```diff
@@ -36,6 +36,117 @@ Payload    Load, verify and launch OS images; or perform firmware update
   If verification fails, one of two things occurs. If resiliency is enabled, the boot is swapped over to the alternate partition
   to recover the broken partition (see :ref:`firmware-resiliency-and-recovery`). Otherwise, the boot is halted.
 
+Stage 1A
+~~~~~~~~
+
+After coming out of reset, Stage 1A sets up the initial exectution environment to support the
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:42]: Please add this back - When coming out of reset, the processor may be in a very raw state and needs some basic
initialization before the rest of the platform can be initialized.

File: source/developer-guides/boot-flow.rst [hunk 16]
```diff
@@ -36,6 +36,117 @@ Payload    Load, verify and launch OS images; or perform firmware update
   If verification fails, one of two things occurs. If resiliency is enabled, the boot is swapped over to the alternate partition
   to recover the broken partition (see :ref:`firmware-resiliency-and-recovery`). Otherwise, the boot is halted.
 
+Stage 1A
+~~~~~~~~
+
```
Comment (rprangar) [inline source/developer-guides/boot-flow.rst:41]: Please add this back - When coming out of reset, the processor may be in a very raw state and needs some basic
initialization before the rest of the platform can be initialized.

**Review by rprangar — APPROVED** [review]
LGTM
